vms/platformvm/deposit/deposit.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2022, Chain4Travel AG. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package deposit
+
+import (
+	"math/big"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/hashing"
+	"github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/avalanchego/vms/platformvm/blocks"
+)
+
+const (
+	interestRateBase        = 365 * 24 * 60 * 60
+	interestRateDenominator = 1_000_000 * interestRateBase
+)
+
+var bigInterestRateDenominator = (&big.Int{}).SetInt64(interestRateDenominator)
+
+type Offer struct {
+	ID ids.ID
+
+	UnlockHalfPeriodDuration uint32 `serialize:"true"`
+	InterestRateNominator    uint64 `serialize:"true"`
+	Start                    uint64 `serialize:"true"`
+	End                      uint64 `serialize:"true"`
+	MinAmount                uint64 `serialize:"true"`
+	MinDuration              uint32 `serialize:"true"`
+	MaxDuration              uint32 `serialize:"true"`
+}
+
+// Sets offer id from its bytes hash
+func (o *Offer) SetID() error {
+	bytes, err := blocks.GenesisCodec.Marshal(blocks.Version, o)
+	if err != nil {
+		return err
+	}
+	o.ID = hashing.ComputeHash256Array(bytes)
+	return nil
+}
+
+// Time when this offer becomes active
+func (o *Offer) StartTime() time.Time {
+	return time.Unix(int64(o.Start), 0)
+}
+
+// Time when this offer becomes inactive
+func (o *Offer) EndTime() time.Time {
+	return time.Unix(int64(o.End), 0)
+}
+
+func (o *Offer) InterestRateFloat64() float64 {
+	return float64(o.InterestRateNominator) / float64(interestRateDenominator)
+}
+
+type Deposit struct {
+	DepositOfferID      ids.ID `serialize:"true"`
+	UnlockedAmount      uint64 `serialize:"true"`
+	ClaimedRewardAmount uint64 `serialize:"true"`
+	Start               uint64 `serialize:"true"`
+	Duration            uint32 `serialize:"true"`
+	Amount              uint64 `serialize:"true"`
+}
+
+func (d *Deposit) StartTime() time.Time {
+	return time.Unix(int64(d.Start), 0)
+}
+
+func (d *Deposit) IsExpired(
+	depositOffer *Offer,
+	timestamp uint64,
+) bool {
+	return d.Start+uint64(d.Duration)+uint64(depositOffer.UnlockHalfPeriodDuration) < timestamp
+}
+
+// Returns amount of tokens that can be unlocked from [deposit] at [unlockTime] (seconds).
+//
+// Precondition: all args are valid in conjunction.
+func UnlockableAmount(deposit *Deposit, offer *Offer, unlockTime uint64) uint64 {
```

Comment:
I would recommend renaming the function to "**Calculate**UnlockableAmount"

Reply:
what do u think on making this deposit method?

Reply:
sorry, I didn't get that